python/demo/demo_poisson_matrix_free.py

```diff
@@ -0,0 +1,508 @@
+bc = fem.dirichletbc(value=uD, dofs=dofs)
+
+# Next, we express the variational problem using UFL.
+
+x = ufl.SpatialCoordinate(msh)
+u = ufl.TrialFunction(V)
+v = ufl.TestFunction(V)
+f = fem.Constant(msh, ScalarType(-6))
+a = inner(grad(u), grad(v)) * dx
+L = inner(f, v) * dx
+
+# For the matrix-free solvers we also define a second linear form `M` as
+# the {py:class}`action <ufl.action>` of the bilinear form $a$ onto an
+# arbitrary {py:class}`Function <dolfinx.fem.Function>` `ui`. This linear
+# form is defined as
+#
+# $$
+# M(v) = a(u_i, v) \quad \text{for} \; \ u_i \in V.
+# $$
+
+ui = fem.Function(V)
+M = action(a, ui)
+
+# ### Direct solver using the assembled matrix
+#
+# To validate the results of the matrix-free solvers, we first compute the
+# solution with a direct solver using the assembled matrix.
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[bc],
+                                  petsc_options={"ksp_type": "preonly",
+                                                 "pc_type": "lu"})
+uh_lu = problem.solve()
+
+
+# The error of the finite element solution `uh_lu` compared to the exact
+# solution $u_{\rm D}$ is calculated below in the $L_2$-norm.
+
+# +
+def L2Norm(u):
+    return np.sqrt(msh.comm.allreduce(
+        fem.assemble_scalar(fem.form(inner(u, u) * dx)),
+        op=MPI.SUM))
+
+
+error_L2_lu = L2Norm(uh_lu - uD)
+if msh.comm.rank == 0:
+    print("Direct solver using the assembled matrix:")
+    print(f"L2-error against exact solution:  {error_L2_lu:.4e}")
+# -
+
+# ### Matrix-free Conjugate Gradient solvers
+#
+# For the matrix-free solvers, the RHS vector $b$ is first assembled based
+# on the linear form $L$.  To account for the Dirichlet boundary conditions
+# in $b$, we apply lifting, i.e. set $b - A x_{\rm bc}$ as new RHS vector $b$.
+# Since we want to avoid assembling the matrix `A`, we compute the necessary
+# matrix-vector product using the linear form `M` implicitly.
+
+b = fem.petsc.assemble_vector(fem.form(L))
+# Apply lifting: b <- b - A * x_bc
+fem.set_bc(ui.x.array, [bc], scale=-1)
+fem.petsc.assemble_vector(b, fem.form(M))
+b.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+fem.petsc.set_bc(b, [bc], scale=0.0)
+b.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+# In the following, different variants are presented in which the posed
+# Poisson problem is solved using matrix-free CG solvers. In each case
+# we want to achieve convergence up to a relative tolerence `rtol = 1e-6`
+# within `max_iter = 200` iterations.
+
+rtol = 1e-6
+max_iter = 200
+
+
+# #### 1. Implementation using PETSc vectors
+
+# To implement the matrix-free CG solver using *PETSc* vectors, we define the
+# function `action_A` with which the matrix-vector product $y = A x$
+# is computed.
+
+def action_A(x):
+    # Update coefficient ui of the linear form M
+    x.copy(ui.vector)
+    ui.vector.ghostUpdate(addv=PETSc.InsertMode.INSERT,
```

Comment:
Can we use `dolfinx` vectors here?
Maybe it's too late for this change because it's already been merged.

Reply:
No, it's not too late :)
I replaced some `ghostUpdate` calls with `scatter_forward` where it is straight-forward and will try using `dolfinx` vectors everywhere later.